spring-batch-core/src/main/java/org/springframework/batch/core/repository/dao/jdbc/JdbcJobInstanceDao.java

```diff
@@ -229,43 +233,32 @@ public List<JobInstance> extractData(ResultSet rs) throws SQLException, DataAcce
 			jobName = jobName.replaceAll("\\" + STAR_WILDCARD, SQL_WILDCARD);
 		}
 
-		return getJdbcTemplate().query(getQuery(FIND_LAST_JOBS_BY_NAME), extractor, jobName);
+		return getJdbcClient().sql(getQuery(FIND_LAST_JOBS_BY_NAME)).param("jobName", jobName).query(extractor);
 	}
 
 	@Override
 	@Nullable
 	public JobInstance getLastJobInstance(String jobName) {
-		try {
-			return getJdbcTemplate().queryForObject(getQuery(FIND_LAST_JOB_INSTANCE_BY_JOB_NAME),
-					new JobInstanceRowMapper(), jobName, jobName);
-		}
-		catch (EmptyResultDataAccessException e) {
-			return null;
-		}
+		return getJdbcClient().sql(getQuery(FIND_LAST_JOB_INSTANCE_BY_JOB_NAME))
+			.param("jobName", jobName)
+			.query(new JobInstanceRowMapper())
+			.optional()
+			.orElse(null);
 	}
 
 	@Override
 	@Nullable
 	public JobInstance getJobInstance(JobExecution jobExecution) {
-
-		try {
-			return getJdbcTemplate().queryForObject(getQuery(GET_JOB_FROM_EXECUTION_ID), new JobInstanceRowMapper(),
-					jobExecution.getId());
-		}
-		catch (EmptyResultDataAccessException e) {
-			return null;
-		}
+		return getJdbcClient().sql(getQuery(GET_JOB_FROM_EXECUTION_ID))
+			.param("jobExecutionId", jobExecution.getId())
+			.query(new JobInstanceRowMapper())
+			.optional()
+			.orElse(null);
 	}
 
 	@Override
 	public long getJobInstanceCount(@Nullable String jobName) throws NoSuchJobException {
-
-		try {
-			return getJdbcTemplate().queryForObject(getQuery(COUNT_JOBS_WITH_NAME), Long.class, jobName);
-		}
-		catch (EmptyResultDataAccessException e) {
-			throw new NoSuchJobException("No job instances were found for job name " + jobName);
-		}
+		return getJdbcClient().sql(getQuery(COUNT_JOBS_WITH_NAME)).param("jobName", jobName).query(Long.class).single();
```

Comment:
Will this throw a `NoSuchJobException` as before if the result set is empty?

Reply:
Actually count statement never generate empty result set.